### ISSUE TITLE
Apply correlated collection optimization even when subquery is streaming

### ISSUE THREAD

**Issue by divega** [dotnet/efcore#11748]
TL;DR: this issue is about changing the design philosophy of EF Core to switch to buffering of query results when facing queries that project nested `IEnumerable<T>` properties, in the interest of reducing the number of database roundtrips. 

See https://github.com/aspnet/EntityFrameworkCore/issues/6953, https://github.com/aspnet/EntityFrameworkCore/issues/5853, and a few others.

### Details

Historically in EF Core and previous versions of EF, most LINQ queries that return collections of objects are evaluated in a streaming fashion. That is, the underlying data readers are consumed on demand as the application consumes the `IEnumerable<T>`. This is the opposite of buffering, which means the underlying data reader (or readers) is fully consumed before the query object starts yielding results to the application. 

One of the theoretical benefits of streaming is that it makes it possible to handle query results that don't fit in memory. However this is not such a common occurrence:

- Queries need to be non-tracking, otherwise the results will have to fit in memory because the DbContext keeps track of every object
- Results that don't fit in memory correlates with the query being long running, which makes the application slow and less reliable (probability of transient failures affecting the query increase)
- You would have to avoid retrying execution strategies, which trigger buffering

Another possible benefit of streaming is that interleaving in-memory work with consuming rows from a reader (e.g. in a for loop) gives more time for the next chunk to be available from the network, decreasing the chances that the data for the next row will not be already available in the buffer, which would trigger blocking (for sync queries) or an "async context switch". This is however very theoretical and we haven’t measured the difference.

On the flip-side, buffering results may lead to closing the connection earlier, which also have potential benefits.

In EF Core, this emphasis on preserving streaming, has lead to situations in which we generate N+1 queries (for example, the two issues linked above). In EF Core 2.1 we have enabled customers to opt out from this behavior by adding calls to ToList() in their nested collections. If we said avoiding N+1 queries is more important that streaming, then this wouldn't be required. 

But what if a user really wanted a query to be resolved in a streaming way? 

https://github.com/aspnet/EntityFrameworkCore/issues/8234 talks about a query operator (e.g. AsStreaming) that can be used to signal this preference.

**Comment by ajcvickers** [comment]
@anpete @smitpatel @maumar to update the title to be clear to the query posse.

**Comment by roji** [comment]
Just to add my 2 cents, avoiding N+1 as described above is very important and could very well justify buffering over streaming in some cases, but I think streaming shouldn't be dismissed too quickly.

> Queries need to be non-tracking, otherwise the results will have to fit in memory because the DbContext keeps track of every object

This is true but I think it's quite a reasonable and expected restriction, which doesn't really make streaming more or less desirable - it's more a question of making sure users are aware via documentation etc.

> Results that don't fit in memory correlates with the query being long running, which makes the application slow and less reliable (probability of transient failures affecting the query increase)

I don't know about this... First, in some non-rare scenarios the application really does have to ingest a lot of data coming from the database - that's just what the user is trying to accomplish. Second, a lot depends on what is done with the data client-side: if processing is quite fast, then it can end up quite similar to buffering in terms of speed.

Of course it's valid to discuss opting out of streaming (via `ToList()`) vs. opting in (via some sort of `AsStreamin()`). As a user, I'd simply expect EF Core to do what's best in general: where N+1 would be generated, buffering would be a reasonable default (but possibly with `AsStreaming()` to override). Wherever streaming can be used without penalty (i.e. no N+1), I'd expect that to be the default. I have no idea how feasible that is though - just expressing expectations.

In general, I just hope you guys avoid penalizing large-dataset scenarios which do occur - I know of some people out there using EF Core for research workloads which are completely non-web, and large datasets do happen. Having a big memory footprint would be a problem in this kind of case.

**Comment by ajcvickers** [comment]
Note: See also #11884, where ToList in the projection results in client-eval in the Where clause. We should decide whether this is desirable behavior given that after fixing this a) much code may have ToList because of the behavior described in this issue and it may be detrimental to force all of it to still client eval and b) ToList means nothing here after making this change, so having as a way to force client eval may be valuable.
